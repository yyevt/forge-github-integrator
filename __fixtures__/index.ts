export * from "./pulls";
export * from "./repos";