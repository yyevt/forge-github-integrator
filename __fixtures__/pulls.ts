import { PullPresentationDto } from "../src/common/PresentationTypes";

export const OPEN_PULL: PullPresentationDto = {
    issueKey: "KAN-1",
    issueIsClosed: false,
    number: 42,
    title: "Add dark mode support",
    url: "https://github.com/org/repo/pull/42",
    state: "OPEN",
    isDraft: false,
    mergeable: "MERGEABLE",
    merged: false,
    repoName: "design-system",
    repoIsDisabled: false,
    repoIsArchived: false,
    authorLogin: "octocat",
};

export const CLOSED_MERGED_PULL: PullPresentationDto = {
    issueKey: "DEF-456",
    issueIsClosed: true,
    number: 17,
    title: "Fix typo in README",
    url: "https://github.com/org/repo/pull/17",
    state: "CLOSED",
    isDraft: false,
    mergeable: "UNKNOWN",
    merged: true,
    repoName: "design-system",
    repoIsDisabled: false,
    repoIsArchived: false,
    authorLogin: "hubot",
};