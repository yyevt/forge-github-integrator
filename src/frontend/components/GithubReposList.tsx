import {Box, EmptyState, Stack} from "@forge/react";
import React from "react";
import {RepoMainInfo} from "./RepoMainInfo";
import {RepoPullsList} from "./RepoPullsList";
import {RepositoryPresentationDto} from "../../common/PresentationTypes";

export const GithubReposList = ({repos}: { repos: ReadonlyArray<RepositoryPresentationDto> }) => {
    return (
        <>
            {repos.length === 0 ? (
                <EmptyState header="Have no repositories"/>
            ) : (
                repos.map(repo => (
                    <Box padding="space.300" backgroundColor="color.background.discovery">
                        <Stack>
                            <RepoMainInfo repo={repo}/>
                            <RepoPullsList
                                pulls={repo.pullRequests}
                                pullsTotalCount={repo.pullRequestsTotalCount}
                                isMergeAllowed={repo.isMergeCommitAllowed}
                            />
                        </Stack>
                    </Box>
                ))
            )}
        </>
    );
}