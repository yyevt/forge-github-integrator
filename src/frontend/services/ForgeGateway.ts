import {invoke} from "@forge/bridge";
import {PullPresentationDto, RepositoryPresentationDto} from "../../common/PresentationTypes";

export class ForgeGateway {

    public async saveToken(token: string): Promise<void> {
        await invoke("saveGithubToken", {token: token});
    }

    public async getGithubReposWithPulls(): Promise<ReadonlyArray<RepositoryPresentationDto>> {
        return invoke("getGithubReposWithPulls");
    }

    public approvePullRequest(pullRequest: PullPresentationDto): Promise<void> {
        return invoke("approvePullRequest", {repo: pullRequest.repoName, pullNumber: pullRequest.number});
    }

    public mergePullRequest(pullRequest: PullPresentationDto): Promise<void> {
        return invoke("mergePullRequest", {repo: pullRequest.repoName, pullNumber: pullRequest.number});
    }

}