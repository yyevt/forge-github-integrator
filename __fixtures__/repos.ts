import {CLOSED_MERGED_PULL, OPEN_PULL} from "./pulls";
import {RepositoryPresentationDto} from "../src/common/PresentationTypes";

export const ACTIVE_REPO: RepositoryPresentationDto = {
    name: "design-system",
    url: "https://github.com/org/design-system",
    ownerLogin: "org",
    languages: ["TypeScript", "CSS"],
    isMergeCommitAllowed: true,
    isDisabled: false,
    isArchived: false,
    visibility: "public",
    createdAt: "2023-05-01T12:00:00Z",
    pullRequestsTotalCount: 2,
    pullRequests: [OPEN_PULL, CLOSED_MERGED_PULL],
};

export const ARCHIVED_REPO: RepositoryPresentationDto = {
    name: "legacy-app",
    url: "https://github.com/org/legacy-app",
    ownerLogin: "org",
    languages: ["JavaScript"],
    isMergeCommitAllowed: false,
    isDisabled: true,
    isArchived: true,
    visibility: "private",
    createdAt: "2020-01-15T08:30:00Z",
    pullRequestsTotalCount: 0,
    pullRequests: [],
};