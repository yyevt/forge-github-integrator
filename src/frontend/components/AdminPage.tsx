import {
    ErrorMessage,
    Form,
    FormFooter,
    FormHeader,
    FormSection,
    HelperMessage,
    Label,
    LoadingButton,
    RequiredAsterisk,
    Textfield,
    useForm,
    ValidMessage
} from "@forge/react";
import React, {useState} from "react";
import {ForgeGateway} from "../services/ForgeGateway";
import {showToast} from "./Utils";

type FormInputs = {
    token: string;
}

export const AdminPage = () => {
    const {register, getFieldId, handleSubmit, formState: {errors, touchedFields}} = useForm<FormInputs>();
    const [isSubmitting, setSubmitting] = useState(false);

    const onSubmit = async (data: FormInputs): Promise<void> => {
        setSubmitting(true);
        try {
            await new ForgeGateway().saveToken(data.token);
            showToast("success", "Success!", "success", "Token was successfully saved.");
        } catch (e) {
            console.error(e);
            showToast("err", "Error", "error", (e instanceof Error ? e.message : `Unknown error occurred: ${e}`));
        } finally {
            setSubmitting(false);
        }
    };

    const isTokenFieldValid = !errors["token"];
    const isTokenFieldTouched = Boolean(touchedFields["token"]);

    return (
        <>
            <Form onSubmit={handleSubmit(onSubmit)}>
                <FormHeader title="Please enter GitHub token">
                    Required fields are marked with an asterisk <RequiredAsterisk/>
                </FormHeader>
                <FormSection>
                    <Label labelFor={getFieldId("token")}>GitHub token <RequiredAsterisk/></Label>
                    <Textfield placeholder="Paste token here" {...register("token", {
                        required: true,
                        pattern: /^(gh[pousr]_[a-zA-Z0-9]{36,255})$/
                    })} />
                    {isTokenFieldValid ? (
                        isTokenFieldTouched ? (
                            <ValidMessage>Valid token format</ValidMessage>
                        ) : (
                            <HelperMessage>You can use classic tokens or fine-grained ones.</HelperMessage>
                        )
                    ) : (
                        <ErrorMessage>Invalid token format</ErrorMessage>
                    )}
                </FormSection>
                <FormFooter>
                    <LoadingButton appearance="primary" type="submit" isLoading={isSubmitting}>
                        Save
                    </LoadingButton>
                </FormFooter>
            </Form>
        </>
    );
}