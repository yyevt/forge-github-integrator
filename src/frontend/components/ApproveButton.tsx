import React, {useCallback, useState} from "react";
import {LoadingButton} from "@forge/react";
import {showToast} from "./Utils";
import {ForgeGateway} from "../services/ForgeGateway";
import {PullPresentationDto} from "../../common/PresentationTypes";

export const ApproveButton = ({pullRequest}: { pullRequest: PullPresentationDto }) => {
    const [isApproving, setIsApproving] = useState(false);

    const approve = useCallback(async (): Promise<void> => {
        setIsApproving(true);
        try {
            await new ForgeGateway().approvePullRequest(pullRequest);
            showToast("success", "Success!", "success", "Pull.ts request was successfully approved.");
        } catch (e) {
            console.error(e);
            showToast("err", "Error", "error", (e instanceof Error ? e.message : `Unknown error occurred: ${e}`));
        } finally {
            setIsApproving(false);
        }
    }, [pullRequest]);

    return (
        <LoadingButton
            isLoading={isApproving}
            onClick={approve}
        >
            Approve
        </LoadingButton>
    );
};