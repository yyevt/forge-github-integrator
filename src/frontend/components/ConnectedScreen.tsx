import React, {useEffect, useState} from "react";
import {SectionMessage, Spinner, Stack, Text} from "@forge/react";
import {ForgeGateway} from "../services/ForgeGateway";
import {GithubReposList} from "./GithubReposList";
import {RepositoryPresentationDto} from "../../common/PresentationTypes";

export const ConnectedScreen = () => {
    const [isLoading, setIsLoading] = useState(false);
    const [repos, setRepos] = useState<ReadonlyArray<RepositoryPresentationDto>>([]);
    const [fetchReposError, setFetchReposError] = useState("");

    useEffect(() => {
        (async () => {
            setFetchReposError("");
            setIsLoading(true);
            try {
                const info = await new ForgeGateway().getGithubReposWithPulls();
                setRepos(info);
            } catch (e) {
                setFetchReposError(e instanceof Error ? e.message : `Unknown error occurred: ${e}`);
            } finally {
                setIsLoading(false);
            }
        })();
    }, []);

    if (isLoading) {
        return <Spinner/>;
    }
    return (
        <Stack space="space.200">
            <SectionMessage appearance="warning">
                <Text>Only limited amount of repositories and pull-requests are shown: max 100 repos, max 100 pulls</Text>
            </SectionMessage>
            {fetchReposError && (
                <SectionMessage appearance="error">
                    <Text>{fetchReposError}</Text>
                </SectionMessage>
            )}
            <GithubReposList repos={repos}/>
        </Stack>
    );
}