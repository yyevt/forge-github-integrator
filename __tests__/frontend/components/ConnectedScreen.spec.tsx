import React from "react";
import renderer, {act, ReactTestRenderer} from "react-test-renderer";
import {ConnectedScreen} from "../../../src/frontend/components/ConnectedScreen";
import {ForgeGateway} from "../../../src/frontend/services/ForgeGateway";
import {ACTIVE_REPO, ARCHIVED_REPO} from "../../../__fixtures__";

jest.mock("@forge/bridge", () => ({
    invoke: jest.fn()
}));

jest.mock("../../../src/frontend/services/ForgeGateway");

describe("ConnectedScreen", () => {

    it("renders screen no repositories correctly", async () => {
        (ForgeGateway.prototype.getGithubReposWithPulls as jest.Mock).mockResolvedValueOnce([]);

        let reactRenderer: ReactTestRenderer;

        await act(async () => {
            reactRenderer = renderer.create(<ConnectedScreen />);
        });

        await act(() => Promise.resolve());

        expect(reactRenderer!.toJSON()).toMatchSnapshot();
    });

    it("renders screen with repositories correctly", async () => {
        (ForgeGateway.prototype.getGithubReposWithPulls as jest.Mock).mockResolvedValueOnce([
            ACTIVE_REPO,
            ARCHIVED_REPO
        ]);

        let reactRenderer: renderer.ReactTestRenderer;

        await act(async () => {
            reactRenderer = renderer.create(<ConnectedScreen />);
        });

        await act(() => Promise.resolve());

        expect(reactRenderer!.toJSON()).toMatchSnapshot();
    });
});