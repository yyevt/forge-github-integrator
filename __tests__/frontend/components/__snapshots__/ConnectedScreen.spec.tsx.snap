// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ConnectedScreen renders screen no repositories correctly 1`] = `
<Stack
  space="space.200"
>
  <SectionMessage
    appearance="warning"
  >
    <Text>
      Only limited amount of repositories and pull-requests are shown: max 100 repos, max 100 pulls
    </Text>
  </SectionMessage>
  <EmptyState
    header="Have no repositories"
  />
</Stack>
`;

exports[`ConnectedScreen renders screen with repositories correctly 1`] = `
<Stack
  space="space.200"
>
  <SectionMessage
    appearance="warning"
  >
    <Text>
      Only limited amount of repositories and pull-requests are shown: max 100 repos, max 100 pulls
    </Text>
  </SectionMessage>
  <Box
    backgroundColor="color.background.discovery"
    padding="space.300"
  >
    <Stack>
      <Heading
        as="h2"
      >
        <Inline>
          <Link
            href="https://github.com/org/design-system"
          >
            design-system
          </Link>
          <Tag
            text="public"
          />
          <Lozenge
            appearance="success"
          >
            Not disabled
          </Lozenge>
          <Lozenge
            appearance="success"
          >
            Not archived
          </Lozenge>
        </Inline>
      </Heading>
      <Text>
        Created: 
        Mon May 01 2023
      </Text>
      <Text>
        Owner: 
        org
      </Text>
      <TagGroup>
        <Tag
          text="TypeScript"
        />
        <Tag
          text="CSS"
        />
      </TagGroup>
      <DynamicTable
        caption="Repository Open PRs: 2"
        rowsPerPage={10}
      >
        <ContentWrapper
          name="head"
        >
          <Cell
            cellKey="issue"
            isSortable={true}
          >
            Issue
          </Cell>
          <Cell
            cellKey="isIssueClosed"
            isSortable={false}
          >
            Is Closed
          </Cell>
          <Cell
            cellKey="title"
            isSortable={true}
          >
            Title
          </Cell>
          <Cell
            cellKey="author"
            isSortable={true}
          >
            Author
          </Cell>
          <Cell
            cellKey="state"
            isSortable={true}
          >
            State
          </Cell>
          <Cell
            cellKey="isDraft"
            isSortable={false}
          >
            Is Draft
          </Cell>
          <Cell
            cellKey="isClean"
            isSortable={false}
          >
            Is Clean
          </Cell>
          <Cell
            cellKey="actions"
            isSortable={false}
          >
            Actions
          </Cell>
        </ContentWrapper>
        <ContentWrapper
          name="rows"
        >
          <Row
            rowKey="row-0-https://github.com/org/repo/pull/42"
          >
            <Cell
              cellKey="issue"
            >
              <Link
                href="/browse/KAN-1"
              >
                KAN-1
              </Link>
            </Cell>
            <Cell
              cellKey="isIssueClosed"
            >
              <Lozenge
                appearance="moved"
              >
                No
              </Lozenge>
            </Cell>
            <Cell
              cellKey="title"
            >
              <Link
                href="https://github.com/org/repo/pull/42"
              >
                Add dark mode support
              </Link>
            </Cell>
            <Cell
              cellKey="author"
            >
              octocat
            </Cell>
            <Cell
              cellKey="state"
            >
              <Lozenge
                appearance="inprogress"
              >
                OPEN
              </Lozenge>
            </Cell>
            <Cell
              cellKey="isDraft"
            >
              <Lozenge
                appearance="moved"
              >
                No
              </Lozenge>
            </Cell>
            <Cell
              cellKey="isClean"
            >
              <Lozenge
                appearance="moved"
              >
                Yes
              </Lozenge>
            </Cell>
            <Cell
              cellKey="actions"
            >
              <ButtonGroup>
                <LoadingButton
                  isLoading={false}
                  onClick={[Function]}
                >
                  Approve
                </LoadingButton>
                <LoadingButton
                  isDisabled={false}
                  isLoading={false}
                  onClick={[Function]}
                >
                  Merge
                </LoadingButton>
              </ButtonGroup>
            </Cell>
          </Row>
          <Row
            rowKey="row-1-https://github.com/org/repo/pull/17"
          >
            <Cell
              cellKey="issue"
            >
              <Link
                href="/browse/DEF-456"
              >
                DEF-456
              </Link>
            </Cell>
            <Cell
              cellKey="isIssueClosed"
            >
              <Lozenge
                appearance="moved"
              >
                Yes
              </Lozenge>
            </Cell>
            <Cell
              cellKey="title"
            >
              <Link
                href="https://github.com/org/repo/pull/17"
              >
                Fix typo in README
              </Link>
            </Cell>
            <Cell
              cellKey="author"
            >
              hubot
            </Cell>
            <Cell
              cellKey="state"
            >
              <Lozenge
                appearance="inprogress"
              >
                CLOSED
              </Lozenge>
            </Cell>
            <Cell
              cellKey="isDraft"
            >
              <Lozenge
                appearance="moved"
              >
                No
              </Lozenge>
            </Cell>
            <Cell
              cellKey="isClean"
            >
              <Lozenge
                appearance="moved"
              >
                No
              </Lozenge>
            </Cell>
            <Cell
              cellKey="actions"
            >
              <ButtonGroup>
                <LoadingButton
                  isLoading={false}
                  onClick={[Function]}
                >
                  Approve
                </LoadingButton>
                <LoadingButton
                  isDisabled={true}
                  isLoading={false}
                  onClick={[Function]}
                >
                  Merge
                </LoadingButton>
              </ButtonGroup>
            </Cell>
          </Row>
        </ContentWrapper>
      </DynamicTable>
    </Stack>
  </Box>
  <Box
    backgroundColor="color.background.discovery"
    padding="space.300"
  >
    <Stack>
      <Heading
        as="h2"
      >
        <Inline>
          <Link
            href="https://github.com/org/legacy-app"
          >
            legacy-app
          </Link>
          <Tag
            text="private"
          />
          <Lozenge
            appearance="removed"
          >
            Disabled
          </Lozenge>
          <Lozenge
            appearance="removed"
          >
            Archived
          </Lozenge>
        </Inline>
      </Heading>
      <Text>
        Created: 
        Wed Jan 15 2020
      </Text>
      <Text>
        Owner: 
        org
      </Text>
      <TagGroup>
        <Tag
          text="JavaScript"
        />
      </TagGroup>
      <EmptyState
        header="Have no opened pull requests"
      />
    </Stack>
  </Box>
</Stack>
`;
